Sources/Core/Microsoft.StreamProcessing/Aggregates/TopKState.cs

```diff
@@ -0,0 +1,236 @@
+            values.AddAll(other.GetSortedValues());
+        }
+
+        public SortedMultiSet<T> GetSortedValues()
+        {
+            return values;
+        }
+
+        public void Remove(T input, long timestamp)
+        {
+            values.Remove(input);
+        }
+
+        public void RemoveAll(ITopKState<T> other)
+        {
+            values.RemoveAll(other.GetSortedValues());
+        }
+    }
+
+    internal class HoppingTopKState<T> : ITopKState<T>
+    {
+        public long currentTimestamp;
+
+        public SortedMultiSet<T> previousValues;
+        public SortedMultiSet<T> currentValues;
+
+        public int k;
+
+        public Comparison<T> rankComparer;
+
+        public HoppingTopKState(int k, Comparison<T> rankComparer, Func<SortedDictionary<T, long>> generator)
+        {
+            this.k = k;
+            this.rankComparer = rankComparer;
+            this.currentValues = new SortedMultiSet<T>(generator);
+            this.previousValues = new SortedMultiSet<T>(generator);
+        }
+
+        public void Add(T input, long timestamp)
+        {
+            if (timestamp > currentTimestamp)
+            {
+                MergeCurrentToPrevious();
+                currentTimestamp = timestamp;
+            }
+            else if (timestamp < currentTimestamp)
+            {
+                throw new ArgumentException("Invalid timestamp");
+            }
+
+            currentValues.Add(input);
+
+            var toRemove = currentValues.TotalCount - k;
+            if (toRemove > 0)
+            {
+                var min = currentValues.GetMinItem();
+                if (toRemove == min.Count)
+                {
+                    currentValues.Remove(min.Item, min.Count);
+                }
+                else if (toRemove > min.Count)
+                {
+                    throw new InvalidOperationException("CurrentValues has more items than required");
+                }
+            }
+        }
+
+        public void Remove(T input, long timestamp)
+        {
+            if (timestamp < currentTimestamp)
+            {
+                previousValues.Remove(input);
+            }
+            else if (timestamp == currentTimestamp)
+            {
+                currentValues.Remove(input);
+            }
+            else
+            {
+                throw new ArgumentException("Invalid timestamp");
+            }
+        }
+
+        public void RemoveAll(ITopKState<T> other)
+        {
+            if (other.Count != 0)
+            {
+                if (other is HoppingTopKState<T> otherTopK)
+                {
+                    if (otherTopK.currentTimestamp >= currentTimestamp)
+                    {
+                        throw new ArgumentException("Cannot remove entries with current or future timestamp");
+                    }
+                    previousValues.RemoveAll(otherTopK.currentValues);
+                    previousValues.RemoveAll(otherTopK.previousValues);
```

Comment:
It looks like you only handle the case where otherTopK.currentTimestamp > currentTimestamp, but allow otherTopK.currentTimestamp == currentTimestamp